syft/frameworks/torch/mpc/primitives.py

```diff
@@ -0,0 +1,160 @@
+from collections import defaultdict
+from typing import List, Tuple, Union
+
+import torch as th
+import syft as sy
+from syft.exceptions import EmptyCryptoPrimitiveStoreError
+from syft.workers.abstract import AbstractWorker
+
+
+class PrimitiveStorage:
+    """
+    Used by normal workers to store crypto primitives
+    Used by crypto providers to build crypto primitives
+    """
+
+    def __init__(self, owner):
+        # The different kinds of primitives available
+        self.fss_eq: list = []
+        self.fss_comp: list = []
+        self.beaver: list = []
+        self.xor_add_couple: list = []  # couple of the same value shared via ^ or + op
+
+        self._owner: AbstractWorker = owner
+        self._builders: dict = {
+            "fss_eq": self.build_fss_keys(type_op="eq"),
+            "fss_comp": self.build_fss_keys(type_op="comp"),
+            "beaver": self.build_triples,
+            "xor_add_couple": self.build_xor_add_couple,
+        }
+
+    def get_keys(self, type_op, n_instances=1, remove=True):
+        """
+        Return FSS keys primitives
+
+        Args:
+            type_op: eq, comp, or xor_add
+            n_instances: how many primitives to retrieve. Comparison is pointwise so this in
+                convenient
+            remove: if true, pop out the primitive. If false, only read it. Read mode is
+                needed because we're working on virtual workers and they need to gather
+                a some point and then re-access the keys.
+        """
+        type_op = {"eq": "fss_eq", "comp": "fss_comp", "xor_add": "xor_add_couple"}[type_op]
+        primitive_stack = getattr(self, type_op)
+
+        available_instances = len(primitive_stack[0]) if len(primitive_stack) > 0 else -1
+        if available_instances >= n_instances:
+            keys = []
+            for i, prim in enumerate(primitive_stack):
+                # We're selecting on the last dimension of the tensor
+                # [:] ~ [slice(None)]
+                # [:1] ~ [slice(1)]
+                # [1:] ~ [slice(1, None)]
+                # [:, :, :1] ~ [slice(None)] * 2 + [slice(1)]
+                n_dim = len(prim.shape)
+                get_slice = [slice(None)] * (n_dim - 1) + [slice(n_instances)]
+                remaining_slice = [slice(None)] * (n_dim - 1) + [slice(n_instances, None)]
+
+                keys.append(prim[get_slice])
+                if remove:
+                    primitive_stack[i] = prim[remaining_slice]
```

Comment:
Could you specify the form of the primitive_stacks somewhere in the docstring?
Because I don't know if it could be simpler or if I just miss something ^^

Reply:
I've added it :)